tests/jballerina-unit-test/src/test/resources/test-src/types/function/type_def_with_function_type.bal

```diff
@@ -0,0 +1,221 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+
+int gVar1 = 10;
+
+type F function(int a, int b, int c) returns int;
+
+type D function(int a, int b = 12, int c = 13) returns int;
+
+type U int|function(int a = 11, int b = 12, int c = 13) returns int;
+
+type I readonly & function(int a = 11, int b = 12, int c = 13) returns int;
+
+type P function(function(int, int b = 10, int c = 15) returns int i, int d = 12, int e = 13) returns int;
+
+type L function(int k, int d = 12, int e = 13) returns function(int a, int b = 10, int c = 15) returns int;
+
+type K record {|
+    function (int a, int b = 10, int c = 14) returns int d;
+|};
+
+int gVar2 = 20;
+
+type M record {|
+    function (int a, int b = 10, int c = 14) returns int d;
+    function (string, string lName) returns string ...;
+
+|};
+
+type N record {|
+    *M;
+|};
+
+type O map<function(int a, int b, int c) returns int>;
+
+type Q function(int a = gVar1, int b = gVar2, int c = gVar3) returns int;
+
+type R function(int a = gVar1, int b = gVar2 + gVar1, int c = gVar3) returns int;
+
+type S function(int a = gVar1, int b = gVar2 + gVar1 + a, int c = gVar3 + a + b) returns int;
+
+function testOnlyWithRequiredParams() returns F {
+    F func = bar;
+    return func;
+}
+
+function testWithDefaultParams() returns int {
+    D func = bar;
+    return func(10);
+}
+
+function testWithDefaultParams2() returns int {
+    D func = bar;
+    return func(10, 20);
+}
+
+function testWithDefaultParams3() {
+    Q func = bar;
+    R func2 = bar;
+    S func3 = bar;
+
+    assertEquality(60, func());
+    assertEquality(70, func2());
+    assertEquality(130, func3());
+
+    assertEquality(70, func(20));
+    assertEquality(70, func2(20,20));
+    assertEquality(110, func3(20, 20));
+    assertEquality(60, func3(20, 20, 20));
+}
+
+function testUnionTypeWithFunctionType() returns int {
+    U func = bar;
+    if !(func is int) {
+        return func();
+    } 
+    return 0;
+}
+
+function testIntersectionTypeWithFunction() returns int {
+    I func = bar;
+    return func();
+}
+
+function testReturnWithFunctionType() returns int {
+    P func = foo;
+    return func(bar);
+}
+
+// function testReturnWithFunctionType2() returns int {
+//     L func = baz;
+//     var value = func(10, 10, 8);
+//     return value(10);
+// }
+
+// function testRecordTypeWithFuncTypeField() returns int {
+//     K func = {d:bar};
+//     int val = func.d(2);
+//     return val;
+// }
+
+function testRecordTypeWithFuncTypeRestField() returns string {
+    M func = {d:bar, "name": mat};
+    var val = func["name"];
+    if !(val is ()) {
+        return val("Chiran", "Sachintha");
+    }
+    return " ";
+}
+
+// function testRecordTypeWithFuncTypeInclusion() returns int {
```

Comment:
why these are commented out ?

Reply:
These cases are not valid with the fix (https://github.com/ballerina-platform/ballerina-lang/issues/31351)